### DISCUSSION TITLE
Why runtimeConfig is not available at runtime?

### DISCUSSION THREAD

**Discussion by groteck** [nuxt/nuxt#16133] [Questions]
I was few days hammering my head with why the runtime config is not updated. Until I found [this](https://v3.nuxtjs.org/guide/going-further/internals#runtime-context-vs-build-context). Then I realize that `runtimeConfig` is more `buldTimeConfig` is only available at `build`/`dev` time.

**Context:**
This is a classic dockerfile optimized to don't rebuild your apps:

```dockerfile
# Install dependencies only when needed
FROM node:14-alpine AS deps
RUN apk add --no-cache libc6-compat
WORKDIR /app
COPY package.json yarn.lock ./
RUN yarn install --frozen-lockfile

# Rebuild the source code only when needed
FROM node:14-alpine AS builder
WORKDIR /app

COPY . .
COPY --from=deps /app/node_modules ./node_modules
RUN yarn build

# Production image, copy all the files and run the server
FROM node:14-alpine AS runner

WORKDIR /app
COPY --from=builder /app .

RUN addgroup -g 1001 -S nodejs
RUN adduser -S nuxt -u 1001

ENV NUXT_HOST=0.0.0.0

USER nuxt

EXPOSE 3000

CMD ["yarn", "start"]
```

and this is the one that I need to use because I didn't find a way to pass configurations with env vars:

```dockerfile
# Install dependencies only when needed
FROM node:14-alpine AS deps
RUN apk add --no-cache libc6-compat
WORKDIR /app
COPY package.json yarn.lock ./
RUN yarn install --frozen-lockfile

# Rebuild the source code only when needed
FROM node:14-alpine
WORKDIR /app

COPY . .
COPY --from=deps /app/node_modules ./node_modules

ENV NUXT_HOST=0.0.0.0

EXPOSE 3000

CMD ["yarn", "start"]
```

Should we update the documentation and maybe the name of the configuration to be clear that this is `build time` in the `nuxt.config` section?
Should we document a way to propagate dynamic configuration after build (if there is any way)?

**Comment by danielroe** [comment]
Runtime config is overrideable at runtime - see https://v3.nuxtjs.org/guide/features/runtime-config#environment-variables.

**Reply by MarioC3** [reply]
Can runtime config be overridden server-side? like this: nuxt/framework#5636 ?

**Reply by danielroe** [reply]
The reason we don't allow it to be overriden on server-side, from _within_ the app, is that the config is shared between requests and you would end up with potential cross-request state pollution. Instead, I would suggest propagating it in a slightly different way. Maybe with setting it on `event.context` (if you only need it on server-side) or using `useState` if you want to persist it to client-side?

**Reply by MarioC3** [reply]
Aww, man! I thought that was going to work! But it looks like `event.context`  does not persist on different routes since it's the specific context for that event. 

Daniel, Does Nuxt have a way to persist state on the server? I made another question (#5682 )

**Comment by vedmant** [comment]
Yes, the documentation is misleading https://nuxt.com/docs/guide/going-further/runtime-config it says "the ability to update it at runtime by setting environment variables" which is not true as you can't update settings at runtime.

It's all just hardcoded on build time like this:

```
// .output/server/chunks/nitro/node-server.mjs
const _runtimeConfig = {"app":{"baseURL":"/","buildAssetsDir":"/_nuxt/","cdnURL":""},"nitro":{"routeRules":{"/__nuxt_error":{"cache":false}},"envPrefix":"NUXT_"},"public":{"clientId":"bla","accountDomain":"bla","logoutRedirectUrl":"bla","siteProtocol":"http","apiBaseURL":"bla","releaseStage":"development","isDev":true},"clientSecret":"bla"};
```
Which means all the logic for defining runtimeConfig from env variables in `nuxt.config.js` is just hardcoded.

**Reply by danielroe** [reply]
What do you mean? This is the case.

**Reply by vedmant** [reply]
@danielroe Example is: 

```js
// nuxt.config.js
...
   runtimeConfig: {
    public: {
      releaseStage: process.env.RELEASE_STAGE || 'production',
    },
  },
```
I have and .env file with "RELEASE_STAGE=development"

Then I use `const conf = useRuntimeConfig()` in a nitro server plugin, which is always returning "development" even if I run `RELEASE_STAGE=staging node .output/server/index.mjs`

Here is reproduction example: https://stackblitz.com/edit/github-o8ojpu?file=nuxt.config.js

run `yarn build` then `RELEASE_STAGE=staging node .output/server/index.mjs`
see in console that `releaseStage` is not `staging`

**Reply by danielroe** [reply]
Environment variables need to match the runtime config structure (because in production your Nuxt config is no longer in play). In this case it should be `NUXT_PUBLIC_RELEASE_STAGE`.

**Reply by vedmant** [reply]
@danielroe Ok, but we have different env variables that we want Nuxt app to use, if we don't want to add custom NUXT_PUBLIC_* and want to reuse env variables we use for all other our infrastructure how we can achieve this?

**Reply by danielroe** [reply]
For runtimeConfig, you need to ensure they are set as I've described. You can do this before launching your server with a bash script if you cannot control the names of the variables.

**Reply by vedmant** [reply]
@danielroe My question is how we can use own existing ENV variables in runtime config instead of forcing to redo our conventions in env variables to NUXT_*? So you mean there is just now way and we must use NUXT_* for runtime config?

**Reply by vedmant** [reply]
@danielroe another example, we have ENV variables that we need to use in runtimeConfig but we can't change their names, like `ELB_AUTH_DEVELOPMENT_CLIENT_ID` for example, how we are supposed to use this variables in the app? It was clean and simple in Nuxt 2, just as I mentioned in my example and it worked perfectly.
